lib/src/main/java/io/ably/lib/types/DecodingContext.java

```diff
@@ -0,0 +1,38 @@
+package io.ably.lib.types;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.nio.charset.StandardCharsets;
+
+public class DecodingContext {
+
+	private String lastMessageString;
+	private byte[] lastMessageBinary;
+
+	public DecodingContext(Map<String, PluggableCodec> codecs)
+	{
+		lastMessageBinary = null;
+		lastMessageString = null;
+		this.codecs = codecs;
+	}
+
+	public Map<String, PluggableCodec> codecs;
+
+	public byte[] get_lastMessage() {
```

Comment:
1. That's a bug. Thanks for pointing it out. Fixed now.

2. We are setting this on each string message the library receives on each channel. We are getting this only if next message is delta. For performance reasons converting to utf-8 is done only when needed.

The reasoning behind this class design (including UTF-8)  is documented [in the PR description](https://github.com/ably/ably-java/pull/532#issue-345383943).

Reply:
ok... could it not be documented a little closer to home - in the code itself?
I guess it could be in the spec (which I know you're working on) but I'm guessing that optimisations like that are going to be specific to particular client library SDK implementation needs.

Reply:
Well that should again be part of our discussion on how we choose which methods to document deeply. There is non-trivial, last mesage bookkeping like this in a couple of more places in the code. Why do we document this one but not the other ones?

Reply:
I just think some documentation is better than none, and documentation is always best written at the point the code it's documenting is written. As an average developer, I had to ask questions around these interfaces which suggests that others would find themselves in the same situation later. Given we have quite a number of people who maintain the codebase now and there will be more in time, we just need to make sure they get as much help as possible when they find themselves back in this place.